### DISCUSSION TITLE
Observing a response is successfully written

### DISCUSSION THREAD

**Discussion by alextes** [tokio-rs/axum#2490] [Q&A] [answered]
### Summary

This seems rather plain but appears complicated 😅.

We run a service where we intentionally reply close to the client side connection timeout. So I'd be happy to either notice when a connection is closed prematurely by the client or the server has sent all it wanted. I dug down to hyper for the former. Seems there is [some 100 line way of doing it over there](https://github.com/hyperium/hyper/issues/707#issuecomment-1581478752) if you happen to be using hyper directly. So then I thought maybe instead of connection closed, which seems complicated for low level reasons, we could detect the last byte is written to lower layers. Having the TCP connection close successfully would be ideal, but something like last body byte written to a socket could be close enough. I have again found nothing but am hopeful someone here knows how to achieve that.

As a minimum, [I found this discussion](https://github.com/tokio-rs/axum/discussions/1094) where a guard is used to detect the invoked handler fn got dropped, you could combine this with a flag set right before the return. That way you could check there whether a response was returned, or the connection closed before. Decent. Curious if anyone sees a way to do better.

I also tried pulling back TraceLayer to see if it could be helpful. It can log chunks of body written. Still not ideal. Or a successful response being passed off by the handler. Still pretty far from "have the bytes been successfully sent to the client".

For context, client based solutions are out. We don't control the client. We get one request, the client will close the connection after a timeout, we want to check if we responded in time.

### axum version

0.6.20

**Comment by davidpdrsn** [comment, marked as answer]
axum intentionally doesn't implement that actual transport layer, that is what hyper is for. So you have to rely on hyper's low-level API to do this.

**Reply by alextes** [reply]
Thank you for the quick reply!

I'm really enjoying Axum for many of the common web framework tasks. Extract a header, require a query, decode some Json, share some state between requests 👏 . Thanks for the hard work 💛 !

Any examples or documentation anywhere on how to hook into the hyper layer whilst taking as little ownership as possible of the lower layers?

**Reply by davidpdrsn** [reply]
I’m not sure. Have a look at hyper.rs and otherwise ask in the hyper repo.

**Reply by alextes** [reply]
Thanks! Looking at hyper now.

And how about extending Axum to be able to do this? We're not looking for axum to implement the transport layer here, merely to hear back from Hyper that yes, those bytes you gave me were successfully received by the client, which hyper certainly is aware of, and doesn't sound unreasonable to me if "send these bytes" is interpreted as "send these bytes successfully" not merely "attempt to send these bytes for me". Common internet protocols are designed for the former after all.

**Reply by davidpdrsn** [reply]
I’m not sure I’d want to build that into axum. I think requiring people to go to hyper is fine.
